### PR TITLE
Vault Refactor & Meta Oracle

### PR DESCRIPTION
**Pull request by aalavandhan**
* New unified pricer / meta oracle (which is used by bill broker and the charm vaults)
* Updated bill broker to use the new pricer, some code clean up
* Updated charm vault managers to use the new pricer, some code clean up
* Updated test suite

### CODE REVIEW

File: spot-vaults/contracts/_utils/UniswapV3PoolHelpers.sol [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+ *
+ */
+library UniswapV3PoolHelpers {
+    /// @notice Calculates the Time-Weighted Average Price (TWAP) given the TWAP tick and unit token amounts.
+    /// @param twapTick The Time-Weighted Average Price tick.
+    /// @param token0UnitAmt The fixed-point amount of token0 equivalent to 1.0.
+    /// @param token1UnitAmt The fixed-point amount of token1 equivalent to 1.0.
+    /// @param one 1.0 represented in the same fixed point denomination as calculated TWAP.
+    /// @return The computed TWAP price.
+    function calculateTwap(
```
Comment (aalavandhan) [inline spot-vaults/contracts/_utils/UniswapV3PoolHelpers.sol:22]: All this comes from existing code
https://github.com/ampleforth/spot/blob/d6e9cf53507b4d76f769fe6b651330c4e055bd3e/spot-vaults/contracts/WethWamplManager.sol#L300

File: spot-vaults/contracts/_utils/UniswapV3PoolHelpers.sol [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        uint256 ratioX192 = uint256(sqrtPriceX96) * sqrtPriceX96;
+        uint256 twapPrice = FullMath.mulDiv(one, (1 << 192), ratioX192);
+        return FullMath.mulDiv(twapPrice, token1UnitAmt, token0UnitAmt);
+    }
+
+    /// @notice Retrieves the Time-Weighted Average Price (TWAP) tick from a Uniswap V3 pool over a given duration.
+    /// @param pool The Uniswap V3 pool.
+    /// @param twapDuration The TWAP duration.
+    /// @return The TWAP tick.
+    function getTwapTick(
```
Comment (aalavandhan) [inline spot-vaults/contracts/_utils/UniswapV3PoolHelpers.sol:38]: This comes from here: 
https://tienshaoku.medium.com/a-guide-on-uniswap-v3-twap-oracle-2aa74a4a97c5
https://github.com/charmfinance/alpha-vaults-contracts/blob/07db2b213315eea8182427be4ea51219003b8c1a/contracts/AlphaStrategy.sol#L136

File: spot-vaults/contracts/BillBroker.sol [hunk 3]
```diff
@@ -931,6 +916,21 @@ contract BillBroker is
     //-----------------------------------------------------------------------------
     // Private methods
 
+    /// @dev Constructs the new reserve state based on provided balances.
+    function _updatedReserveState(
```
[the remark this answers is not in the document]
Comment (aalavandhan) [reply spot-vaults/contracts/BillBroker.sol:920]: keeping it as is as gains aren't significant

**Review by brandoniles — APPROVED** [review]
LGTM

**Pull request merged**